README.md
# MagicMirror<sup>2</sup> Module: MMM-Homematic-Thermostats
A MagicMirror Module for your radio controlled Homematic (eQ-3) radiator thermostats of type: <b><a href="http://amzn.to/2nzhd4a" target="_blank" title="">HM-CC-RT-DN</a></b> and wall thermostats of type <b><a href="https://www.amazon.de/dp/B00H7UIMGA" target="_blank" title="">HM-TC-IT-WM-W-EU</a></b><br>
According documentation, it should also work with devices of type HM-CC-RT-DN-BoM.

from @spitzlberger extended by wall thermostats, style "table", humidity display, warnings, forming of numbers, ... (2019-02-16)

### The module displays the following information:
<table width="100%" border="0">
	<tbody>
		<tr>
			<td>
				<ul>
				<li>The name of the heater</li>
				<li>The current temperature</li>
				<li>The target temperature</li>
				<li>The current humidity (only wall thermostat)</li>
				<li>The current mode</li>
				<li>The fault reporting (i.e. low battery), if any</li>
				</ul>
				Supported languages are: English, Deutsch
			</td>
			<td>
				<img src="https://cloud.githubusercontent.com/assets/26480749/24071459/8d6b75ac-0bd2-11e7-9685-9817bfea7f00.jpg" border="0" height="250px"><br>
				<span style="font-size:8pt;">Picture: &copy; MAF1981, private</span>
			</td>
			<td>
				<img src="https://user-images.githubusercontent.com/38983450/52899716-60a5ce80-31ed-11e9-8455-9ce4d56f6c7d.jpg" border="0" height="250px"><br>
				<span style="font-size:8pt;">Picture: &copy; spitzlbergerj, <a href="https://github.com/spitzlbergerj/homematic-scripts/blob/master/LICENSE">GPL-3.0</a></span>
			</td>
		</tr>
	</tbody>
</table>

## Screenshots

<table width="100%" border="0" style="border:0;">
	<tbody>
		<tr>
			<td>
				Default screen:
				<p><img src="https://cloud.githubusercontent.com/assets/26480749/24048049/ee2f90c2-0b27-11e7-80d8-8f232328f8a1.png" height="150"/></p>
			</td>
			<td>
				Default screen with a fault message:
				<p><img src="https://cloud.githubusercontent.com/assets/26480749/24048628/c636b6e8-0b29-11e7-9a2d-ce471f058b35.png" height="150"/></p>
			</td>
		</tr>
		<tr>
			<td>
Example: Heater in Kid's room is in manually mode:
				<p><img src="https://cloud.githubusercontent.com/assets/26480749/24070241/a1463d8c-0bb9-11e7-9b17-1ec5aaa29f73.png" height="150"/></p>
			</td>
			<td>
Example: Heater in Kid's room is in holiday mode (party mode in terms of HomeMatic):
				<p><img src="https://cloud.githubusercontent.com/assets/26480749/24070845/38182062-0bc5-11e7-9fef-0ef139fc7e24.png" height="150"/></p>
			</td>
		</tr>
		<tr>
			<td>
Screen showing all information: <p><code>showSetTemperature: true</code><br><code>showCurrentMode: true</code><br>
<code>showFaultReporting: true</code></p>
<p><img src="https://cloud.githubusercontent.com/assets/26480749/24070524/6927f9da-0bbf-11e7-815b-b7786317b8d1.png" height="150"/></p>
			</td>
			<td>
Minimalistic screen:<p><code>showSetTemperature: false</code><br> <code>showCurrentMode: false</code><br>
<code>showFaultReporting: false</code></p><p><img src="https://cloud.githubusercontent.com/assets/26480749/24048630/c63a891c-0b29-11e7-9639-1677d08c5781.png" height="150"/></p>
			</td>
		</tr>
		<tr>
			<td colspan="2">
If you prefer it more colorful, just change the stylesheet...<p><img src="https://cloud.githubusercontent.com/assets/26480749/24070539/c0da85f8-0bbf-11e7-85ba-f03917709805.png" height="150"/></p>
			</td>
		</tr>
		<tr>
			<td colspan="2">
In German language:<p><img src="https://cloud.githubusercontent.com/assets/26480749/24070940/0b329e68-0bc7-11e7-9ff2-db339e0ee64e.png" height="150"/></p>
			</td>
		</tr>
		<tr>
			<td colspan="2">
And finally the style "Table" with warnings for temperature and humidity above and below the threshold:<p><img src="https://user-images.githubusercontent.com/38983450/52899709-51bf1c00-31ed-11e9-9410-abd4d4b6ab25.jpg" height="200"/></p>
			</td>
		</tr>
	</tbody>
</table>

## Pre-requisites
The following dependencies are required and must be installed to be able to use this module:
* <a href="https://github.com/MichMich/MagicMirror" target="_blank" title="MagicMirror2">MagigMirror<sup>2</sup></a><br/>
Obviously yes... without the mirror even this module is useless :-) <br>
Requires at least MM version: 2.1.0
* <a href="https://github.com/hobbyquaker/XML-API" target="_blank" title="XML-API for CCU2">XML-API addon</a><br/><b>The XML-API addon must be installed on your Homematic central control unit (CCU1 / CCU2)</b>. 
* <a href="https://github.com/jindw/xmldom" target="_blank" title="xmldom for node.js">xmldom</a><br/>The xmldom DOMParser and XMLSerializer must be installed for node.js
* <a href="https://github.com/ashtuchkin/iconv-lite" target="_blank" title="iconv-lite for node.js">iconv-lite</a><br/>The iconv-lite is required to deal with the correct character encoding <code>iso-8859-1</code> (i.e.: German umlauts ä ü ö) after receiving data via the XML-API.
* <a href="https://momentjs.com" target="_blank" title="moment.js">moment.js</a><br/>The moment.js library is required to parse the dates coming via XML-API (i.e. to display the end date if a device is in vacation mode). Because moment.js is even used by MagicMirror's default modules, you should already have it installed.
* WiFi/Network - Your CCU2 and your Raspberry Pi has to be connected to your local network. The module communicates with the CCU2 by using HTTP GET-request to retrieve the information. Therfore, each device has to be known by the CCU2 as well.

## Installation
In your terminal, go to your MagicMirror's Module folder:
````
cd ~/MagicMirror/modules
````

Clone this repository:
````
git clone https://github.com/spitzlbergerj/MMM-Homematic-Thermostats
````

Configure the module in your `config.js` file as followed.

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:
````javascript
modules: [
{
	module: 'MMM-Homematic-Thermostats',
	header: 'Myhome: Thermostats',
	position: 'top_left', // This can be any of the regions.
	config: {
		ccu2IP: '127.0.0.1',
		xmlapiURL: 'config/xmlapi',
		updateInterval: 120000,
		style: "table",
		setTempInBrackets: false,
		localeStr: 'de-DE',
		warnColor: 'blue',
		devices: [
			{
				id: '1112',
				label: 'Living Room',
				showSetTemperature: false,
				showCurrentMode: true,
				showFaultReporting: true,
				showHumidity: true,
				precisionTemp: 1,
				precisionHum: 0,
				warnTempHigh: true,
				warnTempLow: true,
				warnHumHigh: true,
				warnHumLow: true,
				tempThresholdLow: 23,
				tempThresholdHigh: 24,
				humThresholdLow: 34,
				humThresholdHigh: 70,
			}
			//Add all other devices you want to show
		],
	}
}
]
````

## Configuration options
The following properties can be configured:

<table width="100%">
	<thead>
		<tr>
			<th>Option</th>
			<th width="100%">Description</th>
		</tr>
	</thead>
	<tbody>
		<tr>
			<td><code>ccu2IP</code></td>
			<td><b>Optional</b></code> - The IP address of your HomeMatic central control unit.
				<br/>If not set, the default is: <code>homematic-ccu2</code></td>
		</tr>
		<tr>
			<td><code>xmlapiURL</code></td>
			<td><b>Optional</b></code> - The URL to the XML-API addon on your HomeMatic central control unit. Is appended to <code>ccu2IP</code>.<br/>If not set, the default is: <code>config/xmlapi</code></td>
		</tr>
		<tr>
			<td><code>updateInterval</code></td>
			<td><b>Optional</b></code> - The update interval in milliseconds.<br/>
				If not set, the default is: <code>300000</code> (5 minutes)</td>
		</tr>
		<tr>
			<td><code>style</code></td>
			<td><b>Optional</b></code> - Determines whether the individual thermostats are displayed as separate lines or as rows in a table. Possible values: <code>'lines'</code> or <code>'table'</code> Default is <code>'lines'</code></td>
		</tr>
		<tr>
			<td><code>setTempInBrackets</code></td>
			<td><b>Optional</b></code> - determines whether the set target temperature is to be displayed in brackets. Possible values: <code>true</code> or <code>false</code> Default is <code>true</code></td>
		</tr>
		<tr>
			<td><code>localeStr</code></td>
			<td><b>Optional</b></code> - String for country-specific formatting of numbers. Possible values: see <a href="https://tools.ietf.org/html/rfc5646">Tags for Identifying Languages</a> Default is <code>'de-DE'</code></td>
		</tr>
		<tr>
			<td><code>warnColor</code></td>
			<td><b>Optional</b></code> - sets the warning color. This value only applies if a value is out of threshold and a warning option is set. Possible values: <code>'red'</code>, <code>'green'</code>, <code>'blue'</code>, <code>'yellow'</code>, <code>'white'</code> Default is <code>'red'</code></td>
		</tr>
		<tr>
			<td><code>devices</code></td>
			<td><b>Required</b> - Add all your devices that should appear in the MagicMirror. Each device must include the following properties:
				<table width="100%">
					<thead>
						<tr>
							<th>Option</th>
							<th width="100%">Description</th>
						</tr>
					</thead>
					<tbody>
						<tr>
							<td><code>id</code></td>
							<td>The unique <code>ise_id</code> to identify the device. All ids can be extracted by calling the following URL of the installed XML-API addon: <code>http://ccu2IP/xmlapiURL/devicelist.cgi</code></td>
						</tr>
						<tr>
							<td><code>label</code></td>
							<td>The label for the device (i.e.: Living Room). If not present or empty, the internal device name is shown instead.</td>
						</tr>
						<tr>
							<td><code>showSetTemperature</code></td>
							<td>Whether to show or to hide the target temperature. Default is <code>false</code></td>
						</tr>
						<tr>
							<td><code>showCurrentMode</code></td>
							<td>Whether to show or to hide the current state (i.e.: Heater off). Default is <code>true</code></td>
						</tr>	
						<tr>
							<td><code>showFaultReporting</code></td>
							<td>Whether to show or to hide any faults of the device (i.e.: Low battery warning). Default is <code>true</code></td>
						</tr>	
						<tr>
							<td><code>showHumidity</code> (only available for wall thermostats)</td>
							<td>Whether to show or to hide the humidity. Default is <code>true</code></td>
						</tr>	
						<tr>
							<td><code>precisionTemp</code></td>
							<td>Decimal places for temperature values. Default is <code>2</code></td>
						</tr>	
						<tr>
							<td><code>precisionHum</code></td>
							<td>Decimal places for humidity values. Default is <code>0</code></td>
						</tr>	
						<tr>
							<td><code>warnTempHigh</code></td>
							<td>Determines whether a warning is displayed when the temperature threshold <code>tempThresholdHigh</code> is exceeded (or equal) by displaying the value in <code>warnColor</code>. Default is <code>'false'</code></td>
						</tr>	
						<tr>
							<td><code>warnTempLow</code></td>
							<td>Determines whether a warning is displayed when the temperature falls below or is equal the threshold <code>tempThresholdLow</code> by displaying the value in <code>warnColor</code>. Default is <code>'false'</code></td>
						</tr>	
						<tr>
							<td><code>warnHumHigh</code></td>
							<td>Determines whether a warning is displayed when the humidity threshold <code>humThresholdHigh</code> is exceeded (or equal) by displaying the value in <code>warnColor</code>. Default is <code>'false'</code></td>
						</tr>	
						<tr>
							<td><code>warnHumLow</code></td>
							<td>Determines whether a warning is displayed when the humidity falls below or is equal the threshold <code>humThresholdLow</code> by displaying the value in <code>warnColor</code>. Default is <code>'false'</code></td>
						</tr>	
						<tr>
							<td><code>tempThresholdLow</code></td>
							<td>Temperature lower threshold. Default is <code>5</code></td>
						</tr>	
						<tr>
							<td><code>tempThresholdHigh</code></td>
							<td>Temperature upper threshold. Default is <code>24</code></td>
						</tr>	
						<tr>
							<td><code>humThresholdLow</code></td>
							<td>Humidity lower threshold. Default is <code>35</code></td>
						</tr>	
						<tr>
							<td><code>humThresholdHigh</code></td>
							<td>Humidity upper threshold. Default is <code>60</code></td>
						</tr>	
						</tbody>
				</table>
			</td>
		</tr>
	</tbody>
</table>

## Troubleshooting
Make sure that the XML-API addon is working fine. If you've setup the XML-API addon and your HomeMatic central control unit (CCU2) with default values, you should be able to see a list of all your connected devices by clicking the following URL:<br> <a href="http://homematic-ccu2/config/xmlapi/devicelist.cgi" target="_blank"><http://homematic-ccu2/config/xmlapi/devicelist.cgi></a><br>
Is everything fine, it should look like:<br>
<img src="https://cloud.githubusercontent.com/assets/26480749/24081489/198aa8ea-0cb5-11e7-93fb-dd43a14b1883.png" height="150"/>

## Side notes
<ul>
<li>The requests via the XML-API addon does not require a login! If your HomeMatic control central unit is accessible without special protection via the Internet, this can be a serious security issue!
</li>
<li>The XML-API does not currently support the following information: Temperature offset, window open detection, descaling, keylock and some more parameter</li>
<li>What is the percentage in the mode? Example:<br>
<img src="https://cloud.githubusercontent.com/assets/26480749/24070974/b20444bc-0bc7-11e7-8f7f-88a9cc1f64ef.png" height="30"/><br>
It shows the current valve state of the device.</li>
</ul>

## Next steps
I'm planning to implement my HomeMatic window handles with the module to show the window state (open, closed). The window handles are of type <a href="http://amzn.to/2mCxxjU" target="_blank">HM-Sec-RHS</a>.

## Notice
There is a very good Magic Mirror Module <a href="https://github.com/Sickboy78/MMM-Homematic">MMM-Homematic</a> for displaying individual values (windows open or closed, energy consumption of washing machine, ...) or system variables.</br><br/>
<img src="https://github.com/Sickboy78/MMM-Homematic/blob/master/screenshot.png" height="150"/>

## Further information
* <a href="http://www.homematic.com/" target="_blank">homematic.com</a> - HomeMatic website 
* <a href="https://www.homematic-inside.de/" target="_blank">homematic-inside.de</a> - Best HomeMatic community (blogs, tipps, addons,...)
* <a href="https://www.homematic-forum.de/" target="_blank">homematic-forum.de</a> - Forum to get help
* <a href="https://www.homematic-inside.de/software/download/item/homematic-skript" target="_blank">HomeMatic script documentation / specification</a>
* <a href="https://www.homematic-inside.de/software/addons/item/xmlapi" target="_blank">XML-API section on homematic-inside.de</a>
* <a href="https://wiki.fhem.de/wiki/HM-CC-RT-DN_Funk-Heizk%C3%B6rperthermostat" target="_blank">FHEM wiki for HM-CC-RT-DN</a> - All information about the device type (German only)


CHANGELOG.md
# MMM-Homematic-Thermostats Change Log
All notable changes to this project will be documented in this file.

## [1.1.1] - 
Include required packages

## [1.1.0] - 
Include package.json
Include Grunt File (Lint) - from template

## [1.0.0] - 
First public release


Gruntfile.js
module.exports = function (grunt) {
	require("time-grunt")(grunt);
	grunt.initConfig({
		pkg: grunt.file.readJSON("package.json"),
		eslint: {
			options: {
				configFile: ".eslintrc.json",
				fix: true
			},
			target: ["*.js"]
		},
		stylelint: {
			simple: {
				options: {
					configFile: ".stylelintrc",
					fix: true
				},
				src: ["*.css"]
			}
		},
		jsonlint: {
			main: {
				src: ["package.json", "translations/*.json"],
				options: {
					reporter: "jshint"
				}
			}
		},
		markdownlint: {
			all: {
				options: {
					config: {
						"default": true,
						"line-length": false,
						"blanks-around-headers": false,
						"no-duplicate-header": false,
						"no-inline-html": false,
						"MD010": false,
						"MD001": false,
						"MD031": false,
						"MD040": false,
						"MD002": false,
						"MD029": false,
						"MD041": false,
						"MD032": false,
						"MD036": false,
						"MD037": false,
						"MD009": false,
						"MD018": false,
						"MD012": false,
						"MD026": false,
						"MD038": false
					}
				},
				src: ["README.md", "CHANGELOG.md", "LICENSE.txt"]
			}
		},
		yamllint: {
			all: [".travis.yml"]
		}
	});
	grunt.loadNpmTasks("grunt-eslint");
	grunt.loadNpmTasks("grunt-stylelint");
	grunt.loadNpmTasks("grunt-jsonlint");
	grunt.loadNpmTasks("grunt-yamllint");
	grunt.loadNpmTasks("grunt-markdownlint");
	grunt.registerTask("default", ["eslint", "stylelint", "jsonlint", "markdownlint", "yamllint"]);
};


node_helper.js
var NodeHelper = require("node_helper");
var request = require("request");
var iconv = require("iconv-lite"); //http://stackoverflow.com/questions/12040643/nodejs-encoding-using-request
var xmldom = require("xmldom");

module.exports = NodeHelper.create({
	start: function () {
		console.log(this.name + " helper started...");
	},

	/**
	 * Receives the socket notitication from the module js.
	 * @param notification The notification name
	 * @param payload The array containing all information to process the request
	 */
	socketNotificationReceived: function (notification, payload) {
		//console.log("IN node_helper.js -> socketNotificationReceived: " + notification + " - " + payload);
		var self = this;
		if (notification === "MMM_CCU2_REQUEST") {
			// console.log(notification +" recieved! ");
			var deviceIds = "";
			for (var deviceObject in payload.deviceList) {
				var device = JSON.parse(payload.deviceList[deviceObject]);
				deviceIds = (deviceIds.length === 0) ? device.deviceId : deviceIds + "," + device.deviceId;
			}
			// console.log("Requesting information from the following devices: " + deviceIds);
			var ccu2url = payload.url;
			if (ccu2url.substring(0, 7) !== "http://") { ccu2url = "http://" + payload.url; }
			if (deviceIds.length !== 0) {
				request({
					url: encodeURI(ccu2url + "/state.cgi?device_id=" + deviceIds),
					method: "GET",
					encoding: null,
				}, function (error, response, body) {
					if (!error && response.statusCode == 200) {
						self.sendSocketNotification("MMM_CCU2_RESPONSE", {
							content: self.processXMLResponse(deviceIds, iconv.decode(body, "iso-8859-1"))
						});
					} else {
						console.error("Could not get data from <" + ccu2url + ">! Error code is " + response.statusCode + ": " + error);
					}
				});
			} else {
				//deviceIds are empty... nothting to request from CCU2
			}

		}
	},

	/**
	 * Process the XML response from the CCU2
	 * !! REQUIRES: sudo npm install -g xmldom --save
	 */
	processXMLResponse: function (deviceIds, xml) {
		// console.log("IN: processXMLResponse... " + deviceIds);
		var xmlDocument = new xmldom.DOMParser().parseFromString(xml, "application/xml");
		var requestedIds = deviceIds.split(",");
		var response = [];
		for (var currentId = 0; currentId < requestedIds.length; currentId++) {
			var currentDeviceId = requestedIds[currentId];
			var currendDevice = xmlDocument.getElementsByTagName("device")[currentId];
			//Check the device id given in the config.js
			if (!currendDevice) {
				//device id does not exist
				continue;
			}
			var deviceName = currendDevice.getAttribute("name");
			var channels = currendDevice.getElementsByTagName("channel");
			var channelObjects = [];
			//Loop over all existing channels and consider only channels for further processing that has datapoints available
			for (ch = 0; ch < channels.length; ch++) {
				var channelObject = channels[ch];
				if (channelObject.getElementsByTagName("datapoint").length > 0) {
					channelObjects.push(channelObject);
				}
			}//for channels
			var item = { [currentDeviceId]: this.createResponse(channelObjects, deviceName, currentDeviceId) };
			response.push(item);
		}
		//console.log("OUT: processXMLResponse! ");
		return response;
	},

	/**
	 * Create the response with all datapoints and available parameters
	 */
	createResponse: function (deviceChannelObjects, deviceName, deviceId) {
		var returnValues = [];
		var deviceArray = {};
		deviceArray["deviceId"] = deviceId;
		deviceArray["deviceName"] = deviceName;
		for (var channelObject in deviceChannelObjects) {
			var datapoints = deviceChannelObjects[channelObject].getElementsByTagName("datapoint");
			//Loop over all available datapoints and return a map with all attributes/values of each datapoint
			for (dp = 0; dp < datapoints.length; dp++) {
				var datapointObject = datapoints[dp];
				var datapointType = datapointObject.getAttribute("type");
				var attributesArray = {};
				for (var i = 0; i < datapointObject.attributes.length; i++) {
					var key = datapointObject.attributes[i].name;
					attributesArray[key] = datapointObject.attributes[i].value;
				}
				deviceArray[datapointType] = attributesArray;
			}
		}
		returnValues.push(deviceArray);
		return returnValues;
	}
});

MMM-Homematic-Thermostats.js
Module.register("MMM-Homematic-Thermostats", {

	// Default module config.
	defaults: {
		devices: [],
		ccu2IP: "homematic-ccu2",
		xmlapiURL: "config/xmlapi",
		updateInterval: 300000,  // 5min
		warnColor: "red",
		showSetTemperature: false,
		showFaultReporting: false,
		showCurrentMode: false,
		showHumidity: false,
		precisionTemp: 2,
		precisionHum: 0,
		warnTempHigh: false,
		warnTempLow: false,
		warnHumHigh: false,
		warnHumLow: false,
		tempThresholdHigh: 24,
		tempThresholdLow: 5,
		humThresholdHigh: 60,
		humThresholdLow: 35,
	},

	homematicUrl: "",
	moduleDisplay: "",
	configurationSettings: [],

	// Override dom generator.
	getDom: function () {
		var wrapper = document.createElement("div");
		wrapper.innerHTML = this.moduleDisplay;
		return wrapper;
	},

	// Define start sequence.
	start: function () {
		var self = this;
		self.homematicUrl = this.config.ccu2IP + "/" + this.config.xmlapiURL;
		self.configurationSettings = self.readConfiguration(this.config);
		setInterval(function () {
			self.getDataFromCCU2();
			self.updateDom();
		}, this.config.updateInterval);
		moment.locale(this.config.language);
		self.getDataFromCCU2();
		Log.info("Starting module: " + this.name + " accessing URL " + self.homematicUrl);
	},

	/**
	 * Send the notification request to the node_helper to get
	 * all settings of all requested devices
	 */
	getDataFromCCU2: function () {
		var self = this;
		this.sendSocketNotification(
			"MMM_CCU2_REQUEST",
			{
				deviceList: self.createPayloadForRequest(),
				url: self.homematicUrl,
			}
		);
	},

	/**
	 * Prepare the settings specified in the config.js
	 * and do some basic checks to prevent common errors
	 */
	readConfiguration: function (config) {
		var deviceSettings = {};

		config.devices.forEach(device => {
			let settings = {
				configDeviceId: device.id,
				configDeviceLabel: device.label || ""
			};

			// RT & WT
			settings["configDeviceShowSetTemperature"] = !!device.showSetTemperature ? device.showSetTemperature : config.showSetTemperature;
			settings["configDeviceShowFaultReport"] = !!device.showFaultReporting ? device.showFaultReporting : config.showFaultReporting;
			settings["configDeviceShowCurrentMode"] = !!device.showCurrentMode ? device.showCurrentMode : config.showCurrentMode;
			settings["configDevicePrecisionTemp"] = !!device.precisionTemp ? device.precisionTemp : config.precisionTemp;
			settings["configDeviceWarnTempHigh"] = !!device.warnTempHigh ? device.warnTempHigh : config.warnTempHigh;
			settings["configDeviceWarnTempLow"] = !!device.warnTempLow ? device.warnTempLow : config.warnTempLow;
			settings["configDeviceTempThresholdHigh"] = !!device.tempThresholdHigh ? device.tempThresholdHigh : config.tempThresholdHigh;
			settings["configDeviceTempThresholdLow"] = !!device.tempThresholdLow ? device.tempThresholdLow : config.tempThresholdLow;

			// WT
			settings["configDeviceShowHumidity"] = !!device.showHumidity ? device.showHumidity : config.showHumidity;
			settings["configDevicePrecisionHum"] = !!device.precisionHum ? device.precisionHum : config.precisionHum;
			settings["configDeviceWarnHumHigh"] = !!device.warnHumHigh ? device.warnHumHigh : config.warnHumHigh;
			settings["configDeviceWarnHumLow"] = !!device.warnHumLow ? device.warnHumLow : config.warnHumLow;
			settings["configDeviceHumThresholdHigh"] = !!device.humThresholdHigh ? device.humThresholdHigh : config.humThresholdHigh;
			settings["configDeviceHumThresholdLow"] = !!device.humThresholdLow ? device.humThresholdLow : config.humThresholdLow;

			deviceSettings[device.id] = settings;
		});

		deviceSettings["global"] = {
			configColumnActTemp: true,
			configColumnSetTemp: config.devices.some(device => !!device.showSetTemperature) || config.showSetTemperature,
			configColumnHumidity: false,
			configColumnMode: config.devices.some(device => !!device.showCurrentMode) || config.showCurrentMode,
		}

		return deviceSettings;
	},

	/**
	 * Creates the payload for the socket notification
	 */
	createPayloadForRequest: function () {
		var payload = [];
		this.config.devices.forEach(device => {
			payload.push(JSON.stringify({ "deviceId": device.id }));
		})
		return payload;
	},

	/**
	 * Receives the notification with the response from the node_helper
	 */
	socketNotificationReceived: function (notification, payload) {
		if (notification === "MMM_CCU2_RESPONSE") {
			if (payload && payload.content) {
				this.moduleDisplay = this.prepareOutputForDevices(payload.content);
				this.updateDom();
			}
		}
	},

	/**
	 * Prepares the output for displaying the values in the mirror.
	 */
	prepareOutputForDevices: function (response) {
		let htmlOutput = "";

		response.forEach(deviceData => {
			for (let deviceId in deviceData) {
				const device = deviceData[deviceId][0]
				const deviceType = !!device["VALVE_STATE"] ? 'RT' : 'WT';

				const configDeviceLabel = this.configurationSettings[deviceId]["configDeviceLabel"]
				const deviceLabel = configDeviceLabel !== "" ? configDeviceLabel : device.deviceName
				const actualTemperature = device["ACTUAL_TEMPERATURE"]["value"];
				const actualTemperatureStr = this.prepareAttribute("ACTUAL_TEMPERATURE", device, this.configurationSettings[deviceId]["configDevicePrecisionTemp"]);
				const setTemperature = this.prepareAttribute("SET_TEMPERATURE", device, this.configurationSettings[deviceId]["configDevicePrecisionTemp"]);
				const currentMode = this.prepareControlModeOutput(device);
				const actualHumidity = deviceType === 'WT' ? device["ACTUAL_HUMIDITY"]["value"] : undefined;
				const actualHumidityStr = deviceType === 'WT' ? this.prepareAttribute("ACTUAL_HUMIDITY", device, this.configurationSettings[deviceId]["configDevicePrecisionHum"]) : undefined;
				const faultMode = this.prepareFaultReporting(device["FAULT_REPORTING"]);

				let classNameStr = ""
				let htmlDeviceEntry = "<span class='deviceContainer'>";

				// Label
				htmlDeviceEntry += "<span class='deviceLabel'>" + deviceLabel + ":&nbsp;</span>"

				// Actual Temp
				classNameStr = "deviceActualTemperature";
				if ((this.configurationSettings[deviceId]["configDeviceWarnTempLow"] && actualTemperature <= this.configurationSettings[deviceId]["configDeviceTempThresholdLow"]) ||
					(this.configurationSettings[deviceId]["configDeviceWarnTempHigh"] && actualTemperature >= this.configurationSettings[deviceId]["configDeviceTempThresholdHigh"])) {
					classNameStr = classNameStr + " bright " + this.config.warnColor;
				}
				htmlDeviceEntry += "<span class='" + classNameStr + "'>" + actualTemperatureStr + "</span>";

				// Set Temp
				if (this.configurationSettings[deviceId]["configDeviceShowSetTemperature"]) {
					htmlDeviceEntry += "<span class='deviceLabel'>&nbsp;/&nbsp;</span><span class='deviceSetTemperature'>" + setTemperature + "</span>"
				}

				// Hum
				if (this.configurationSettings[deviceId]["configDeviceShowHumidity"] && deviceType == "WT") {
					classNameStr = "";
					if ((this.configurationSettings[deviceId]["configDeviceWarnHumLow"] && actualHumidity <= this.configurationSettings[deviceId]["configDeviceHumThresholdLow"]) ||
						(this.configurationSettings[deviceId]["configDeviceWarnHumHigh"] && actualHumidity >= this.configurationSettings[deviceId]["configDeviceHumThresholdHigh"])) {
						classNameStr = " class='bright " + this.config.warnColor + "'";
					}
					htmlDeviceEntry += "<span class='deviceHumidity'>&nbsp;(" + this.translate("UNIT_HUM") + ":&nbsp;" + "<span" + classNameStr + ">" + actualHumidityStr + "</span>)&nbsp;</span>";
				}

				// Mode
				if (this.configurationSettings[deviceId]["configDeviceShowCurrentMode"]) {
					htmlDeviceEntry += "<span class='deviceMode'>&nbsp;(" + currentMode + ")</span>";
				}

				htmlDeviceEntry += "</span>";

				// Show Fault
				if (this.configurationSettings[deviceId]["configDeviceShowFaultReport"]) {
					htmlDeviceEntry += faultMode
				}

				htmlOutput += htmlDeviceEntry;

			}
		})
		return htmlOutput;
	},

	/**
	 * Prepare the output of the given attribute. Reads the attributeName from the
	 * settingsArray and do further processing on it, i.e. to display the value with the
	 * unit (temperature, valve state) or anything else.
	 * Can be used in the future to prepare any other attributes for output.
	 */
	prepareAttribute: function (attributeName, settingsArray, precision) {
		var preparedAttributeValue = "";
		var attributeNameArray = settingsArray[attributeName];
		switch (attributeName) {
			//As of now, all attributes below can be handled with the same logic
			case "ACTUAL_TEMPERATURE":
			case "SET_TEMPERATURE":
			case "VALVE_STATE":
			case "ACTUAL_HUMIDITY":
				preparedAttributeValue = Number(parseFloat(attributeNameArray["value"])).toLocaleString(this.config.localeStr, { minimumFractionDigits: precision, maximumFractionDigits: precision }) + attributeNameArray["valueunit"];
				break;
		}
		return preparedAttributeValue;
	},

	/**
	 * Prepare the control mode of the radiator specified in CONTROL_MODE
	 * parameter. Translate the returning number according to the Homematic
	 * documentation into the available values.
	 */
	prepareControlModeOutput: function (settingsArray) {
		var controlMode = Number(settingsArray["CONTROL_MODE"]["value"]);
		// @spitzlbergerj, 20190210: wall thermostat doesn't contain element VALVE_STATE
		if (settingsArray["VALVE_STATE"]) {
			// @spitzlbergerj, 20190210: Array with element VALVE_STATE, should be radiator thermostat
			var valveState = Number(settingsArray["VALVE_STATE"]["value"]);
			var valveStateDisplay = valveState + settingsArray["VALVE_STATE"]["valueunit"];
		} else {
			var valveState = -1;
			var valveStateDisplay = "-";
		}
		var translatedMode = this.translate("RADIATOR_OFF");
		var modus = "AUTO"; //default
		switch (controlMode) {
			case 0:
				modus = "AUTO";
				if (valveState > 0) {
					//Do not show word "auto" in auto mode: this.translate("RADIATOR_MODE_".concat(modus))
					translatedMode = this.translate("HEATS_WITH") + " " + valveStateDisplay;
				} else if (valveState < 0) {
					//it's a wall thermostat
					translatedMode = this.translate("RADIATOR_MODE") + " " + this.translate("RADIATOR_MODE_".concat(modus));
				}
				break;
			case 1:
				modus = "MANUAL";
				if (valveState !== 0) {
					translatedMode = this.translate("RADIATOR_MODE_".concat(modus)) + ", " + this.translate("HEATS_WITH") + " " + valveStateDisplay;
				} else if (valveState < 0) {
					//it's a wall thermostat
					translatedMode = this.translate("RADIATOR_MODE") + " " + this.translate("RADIATOR_MODE_".concat(modus));
				}
				break;
			case 2:
				modus = "PARTY"; //Urlaubsmodus
				var urlaubsEnde = moment().set({
					"year": settingsArray["PARTY_STOP_YEAR"]["value"],
					"month": (settingsArray["PARTY_STOP_MONTH"]["value"] - 1),
					"date": settingsArray["PARTY_STOP_DAY"]["value"]
				});
				translatedMode = this.translate("RADIATOR_MODE_".concat(modus)) + " " + this.translate("HOLIDAY_MODE_UNTIL") + " ";
				translatedMode = translatedMode + urlaubsEnde.format("ddd, MMM Do Y");
				break;
			case 3:
				modus = "BOOST";
				translatedMode = this.translate("RADIATOR_MODE_".concat(modus));
		}//switch
		return translatedMode;
	},

	/**
	 * Translate the values of parameter CONTROL_MODE to their
	 * human readable values that can be used for i18n afterwards
	 * According Homematic documentation for datapoints
	 */
	prepareControlMode: function (controlModeParameter) {
		var controlMode = Number(controlModeParameter["value"]);
		var modus = "AUTO"; //default
		switch (controlMode) {
			case 0:
				modus = "AUTO";
				break;
			case 1:
				modus = "MANUAL";
				break;
			case 2:
				modus = "PARTY"; //Urlaubsmodus
				break;
			case 3:
				modus = "BOOST";
		}//switch
		return modus;
	},

	/**
	 * Prepare the error (if any) for the FAULT_REPORTING parameter.
	 * Translate the states into their values (according Homematic documentation)
	 * to be able to use it in the translation files. Return an empty
	 * string if everything is fine.
	 */
	prepareFaultReporting: function (faultReporting) {
		let errorMsg = "";
		// @spitzlbergerj, 20190210: wall thermostat doesn't contain element FAULT_REPORTING
		if (!!faultReporting) {
			// @spitzlbergerj, 20190210: Array with element FAULT-REPORTING, should be radiator thermostat
			const faultCode = Number(faultReporting["value"]);
			switch (faultCode) {
				case 1:
					errorMsg = this.translate("VALVE_TIGHT");
					break;
				case 2:
					errorMsg = this.translate("ADJUSTING_RANGE_TOO_LARGE");
					break;
				case 3:
					errorMsg = this.translate("ADJUSTING_RANGE_TOO_SMALL");
					break;
				case 4:
					errorMsg = this.translate("COMMUNICATION_ERROR");
					break;
				case 6:
					errorMsg = this.translate("LOWBAT");
					break;
				case 7:
					errorMsg = this.translate("VALVE_ERROR_POSITION");
					break;
			}
		}

		if (errorMsg !== "") {
			return "<span class='faultReporting'>" + this.translate("WARNING") + errorMsg + "</span>";
		}
		return "";
	},

	// Define required style-sheet scripts
	getStyles: function () {
		return ["MMM-Homematic-Thermostats.css"];
	},

	// Define required dependency scripts
	getScripts: function () {
		return ["moment.js"];
	},

	// Define required translation files
	getTranslations: function () {
		return {
			en: "translations/en.json",
			de: "translations/de.json"
		};
	}
});
